### PR TITLE
Reinforcement Learning using Graph Neural Networks - Practical Course SS20

### PR DESCRIPTION
**Pull request by mrcoliva**
This pull request contains the code of the "Reinforcement Learning using Graph Neural Networks" project.
The main contribution is a GNN-based SAC agent using a graph-based observer.

To train it, execute `bazel run //examples:tfa_gnn -- --mode=train`.

### CODE REVIEW

File: bark_ml/observers/graph_observer.py [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+  This observer converts an `ObservsedWorld` instance into
+  a graph-structured observation, consisting of nodes (with
+  features), an adjacency matrix expressing the (directed)
+  connections between nodes in the graph, and edges (with
+  features), expressing the relative information between
+  the agents that constitute these connections.
+  """
+
+  def __init__(self,
+               normalize_observations=True,
```
Comment (patrickhart) [inline bark_ml/observers/graph_observer.py:32]: Everything else is configured using the `ParameterServer`, should this not also be the case for the normalization?
Would make it easier to store all parameters.

File: bark_ml/observers/graph_observer.py [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+    requested_node_attribute_keys =\
+      params["ML"]["GraphObserver"]["EnabledNodeFeatures",
+      "The list of available node features, given by their string key that \
+       the observer should extract from the world and insert into the \
+       observation. For a list of available features, refer to the list \
+       returned by `GraphObserver.available_node_attributes`.",
+      self.available_node_attributes()
+    ]
+
+    self.enabled_node_attribute_keys = self._filter_requested_attributes(
```
Comment (patrickhart) [inline bark_ml/observers/graph_observer.py:72]: why do we not use an underscore for this and the following member variables?

File: bark_ml/observers/graph_observer.py [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+    obs = np.zeros(self._len_state)
+
+    # insert node features for each agent
+    for i, agent in agents:
+      start_index = i * self.feature_len
+      end_index = start_index + self.feature_len
+      obs[start_index:end_index] = self._extract_node_features(agent)
+
+    edge_features = np.zeros((self._agent_limit, self._agent_limit, self.edge_feature_len))
+    adjacency_matrix = np.zeros((self._agent_limit, self._agent_limit))
```
Comment (patrickhart) [inline bark_ml/observers/graph_observer.py:117]: does this imply the adjacency_matrix only is e.g. 4x4 (as this is set as the agent limit)?
If there are 4 agents, each agent can be connected with 3 other agents assuming directed edges + self connections.
So should the adjacency_matrix not be of larger size (and of course the edge features)?

File: bark_ml/observers/graph_observer.py [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+    return F, A, E
+
+  def _preprocess_agents(self, world):
+    """
+    Preproccesses the agents for constructing an
+    observation by resorting and pruning the 
+    original list of agents.
+
+    Args:
+      world: A `bark.core.World` instance.
```
Comment (patrickhart) [inline bark_ml/observers/graph_observer.py:252]: world: A `bark.core.World` instance.

--> has to be `bark.core.ObservedWorld`

File: bark_ml/observers/graph_observer.py [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+      A list of tuples, consisting of an index and an 
+      agent object element.
+
+      The first element always represents the ego agent.
+      The remaining elements resemble other agents, up 
+      to the limit defined by `self._agent_limit`, sorted 
+      in ascending order with respect to the agents'
+      distance in the world to the ego agent.
+    """
+    ego_agent = world.ego_agent
```
Comment (patrickhart) [inline bark_ml/observers/graph_observer.py:264]: we only have the ego_agent available in the observed world

File: bark_ml/observers/graph_observer.py [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+  @classmethod
+  def available_edge_attributes(cls, with_descriptions=False):
+    attributes = {
+      "dx": "The difference in the x-position of the two agents.", 
+      "dy": "The difference in the y-position of the two agents.", 
+      "dvel": "The difference in the velocity of the two agents.", 
+      "dtheta": "The difference in the heading angle of the two agents."
+    }
+
+    if not with_descriptions:
```
Comment (patrickhart) [inline bark_ml/observers/graph_observer.py:516]: cool!

File: bark_ml/observers/graph_observer.py [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+    the x-element of positions. A normalized value of 1.0 
+    corresponds to the maximum of this range (and vice versa).
+
+    Note: This dictionary does not include a range for each
+    possible attribute, but rather for each _kind_ of attribute,
+    like distances, velocities, angles, etc.
+    E.g. all distances (between agents, between objects, etc.)
+    are scaled relative to the 'distance' range. 
+    """
+    d = OrderedDict()
```
Comment (patrickhart) [inline bark_ml/observers/graph_observer.py:470]: what is the advantage of using an ordered dict?

File: bark_ml/library_wrappers/lib_tf_agents/agents/graph_sac_agent.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    self._eval_policy = self.GetEvalPolicy()
+
+  def GetAgent(self, env, params):
+    gnn_sac_params = self._params["ML"]["BehaviorGraphSACAgent"]
+
+    # actor network
+    actor_net = GNNActorNetwork(
+      input_tensor_spec=env.observation_spec(),
+      output_tensor_spec=env.action_spec(),
+      gnn=GNNWrapper(
```
[the remark this answers is not in the document]
Comment (mrcoliva) [reply bark_ml/library_wrappers/lib_tf_agents/agents/graph_sac_agent.py:61]: Yes, absolutely. Also, using different configurations for the actor and critic GNN would be nice to try. Luckily, both is very easy to implement.

File: bark_ml/library_wrappers/lib_tf_agents/networks/gnn_actor_network.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    
+    flat_action_spec = tf.nest.flatten(output_tensor_spec)
+    
+    if len(flat_action_spec) > 1:
+      raise ValueError('Only a single action is supported by this network')
+    
+    if flat_action_spec[0].dtype not in [tf.float32, tf.float64]:
+      raise ValueError('Only float actions are supported by this network.')
+    
+    if gnn is None:
```
[the remark this answers is not in the document]
Comment (mrcoliva) [reply bark_ml/library_wrappers/lib_tf_agents/networks/gnn_actor_network.py:80]: But do we really require it to be a `GNNWrapper` instance? Every class that has a `call` and `num_units` interface could be used. That was part of the idea why I inject the GNNs from the outside, to be more flexible to simply using another GNN class.

File: bark_ml/library_wrappers/lib_tf_agents/networks/gnn_actor_network.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+      raise ValueError('`gnn` must not be `None`.')
+
+    self._gnn = gnn
+    
+    self._encoder = encoding_network.EncodingNetwork(
+      input_tensor_spec=tf.TensorSpec([None, self._gnn.num_units]),
```
[the remark this answers is not in the document]
Comment (mrcoliva) [reply bark_ml/library_wrappers/lib_tf_agents/networks/gnn_actor_network.py:86]: Yes.

File: bark_ml/observers/graph_observer.py (hunk 1 above)
Comment (mrcoliva) [reply bark_ml/observers/graph_observer.py:32]: Good point. I will change this.

File: bark_ml/observers/graph_observer.py (hunk 7 above)
Comment (mrcoliva) [reply bark_ml/observers/graph_observer.py:470]: actually, there is none.

File: bark_ml/observers/graph_observer.py (hunk 2 above)
Comment (mrcoliva) [reply bark_ml/observers/graph_observer.py:72]: I could imagine that the properties about which features are enabled might be useful to be read from the outside. But in this case, a read-only property might actually be better. I will add an underscore.

File: bark_ml/observers/graph_observer.py (hunk 3 above)
Comment (mrcoliva) [reply bark_ml/observers/graph_observer.py:117]: > does this imply the adjacency_matrix only is e.g. 4x4 (as this is set as the agent limit)?
If there are 4 agents, each agent can be connected with 3 other agents assuming directed edges + self connections.
Yes, that's correct.

> So should the adjacency_matrix not be of larger size (and of course the edge features)?
What makes you think that? 
Assuming 4 agents, the matrix contains a row for each agent (-> 4 rows) and each row contains a binary column for each agent (-> 4 columns) indicating whether an edge to this agent exists. Hence, a 4x4 matrix.

File: bark_ml/observers/graph_observer.py (hunk 5 above)
Comment (mrcoliva) [reply bark_ml/observers/graph_observer.py:264]: I'm not sure I understand, could you please elaborate?

File: bark_ml/library_wrappers/lib_tf_agents/networks/gnn_actor_network.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mrcoliva) [reply bark_ml/library_wrappers/lib_tf_agents/networks/gnn_actor_network.py:80]: Setting the default to `None` would increase the risk of someone missing the argument, right? I added the `gnn is None` check to catch someone explicitly setting the argument to `None` in which case the actor wouldn't work.

[the remark this answers is not in the document]
Comment (patrickhart) [reply bark_ml/library_wrappers/lib_tf_agents/networks/gnn_actor_network.py:80]: Alright, sounds plausible!

File: bark_ml/observers/graph_observer.py (hunk 2 above)
Comment (patrickhart) [reply bark_ml/observers/graph_observer.py]: as you think! was just a question!

File: bark_ml/observers/graph_observer.py (hunk 3 above)
Comment (patrickhart) [reply bark_ml/observers/graph_observer.py:118]: ah okay, well was quite late yesterday!

File: bark_ml/observers/graph_observer.py (hunk 5 above)
Comment (patrickhart) [reply bark_ml/observers/graph_observer.py:265]: This was referring to the comment above. I have seen this before. all good!

File: bark_ml/observers/graph_observer.py (hunk 7 above)
Comment (patrickhart) [reply bark_ml/observers/graph_observer.py:470]: that's what I wanted to know. Just curious; )

Comment (mrcoliva) [reply bark_ml/observers/graph_observer.py]: I removed it for simplicity :)

**Pull request merged**